GitHub pr report

kghamilton89 opened the pull request:
* **GitHub PR Report** setup article;
* Table of Contents **README.md** updated;
* Default credentials information included in other GitHub articles;
* Dictionary updated to include: GraphQL, GQL, OAuth;
* Minor fixes to other GitHub setup articles.

rodionos commented on how-to/github/fork-notification.md:
```diff
@@ -12,11 +12,11 @@ Github is a great place for open-source projects, companies and individuals arou
 ## Launch ATSD Sandbox
 
-Execute the `docker run` command below to launch a local ATSD [sandbox](https://github.com/axibase/dockers/tree/atsd-sandbox) instance.
+Execute the `docker run` command below to launch a local ATSD [sandbox](https://github.com/axibase/dockers/tree/atsd-sandbox) instance. Log in to ATSD using the [default credentials](https://github.com/axibase/dockers/tree/atsd-sandbox#default-credentials).
```
Is it necessary to login into ATSD at this stage? We're instructing the user to launch container with docker run, and then to login, however login is not necessary since it's all automated.

kghamilton89 replied:
Added an **Explore ATSD** section at the end of the article that mentions how to find the ATSD host in the logs and how to log in with the default credentials.

rodionos commented on how-to/github/issue-notification.md:
```diff
@@ -12,13 +12,13 @@ The issue functionality in GitHub has evolved recently to resemble a lightweight
 While the default email notifications delivered by GitHub provide a convenient way to stay on track, the flexibility of being able to match new issues to specific collaborators can be better accomplished using programmatic integration leveraging GitHub webhook functionality.
 
-Webhook functionality is constantly evolving to include new options , explore the GitHub development team's latest pursuits on the [Platform Roadmap](https://developer.github.com/early-access/platform-roadmap/) for an insight into coming features.
+Webhook functionality is constantly evolving to include new options, explore the GitHub development team's latest pursuits on the [Platform Roadmap](https://developer.github.com/early-access/platform-roadmap/) for an insight into coming features.
```
> Webhook functionality is constantly evolving to include new options
I would rephrase this. Webhook functionality is not just evolving (other features are evolving too). GitHub webhook functionality is prominently present on the roadmap, 6 items out of total 13 or 14? Also, it's GitHub's webhook feartures are being upgraded, not ours. 'Pursuit'is a bit too adventurous term for planned work.

kghamilton89 replied:
Rephrased.

rodionos commented on how-to/github/pr-report.md:
```diff
@@ -0,0 +1,173 @@
+
+GitHub features email notifications for individual repositories, but the task of tracking Pull Requests across many repositories may be better accomplished using programmatic integration leveraging the [GraphQL](https://graphql.org/) API query language, featured in the GitHub API.
+
+In contrast to the [GitHub v3 REST API](https://developer.github.com/v3/), the latest API offers more flexibility by replacing multiple REST requests with a single call to fetch all relevant data.
```
`the latest GraphQL API` (add link)

kghamilton89 replied:
Link added.

rodionos commented on how-to/github/pr-report.md:
````diff
@@ -0,0 +1,173 @@
+Replace the `SERVER_URL` parameter with the public DNS name of the Docker host where the sandbox container will be running.
+
+```sh
+docker run -d -p 8443:8443 \
````
In the docker command: a) use only one ATSD_IMPORT_PATH variable (not 3). b) keep only required parameters in the mail.properties file (minimal set of parameters).

kghamilton89 replied:
docker command simplified and extraneous mail.properties parameters removed.
```txt
auth=true
ssl=true
upgrade_ssl=true
```

rodionos commented on how-to/github/pr-report.md:
````diff
@@ -0,0 +1,173 @@
+upgrade_ssl=true
+```
+
+This file defines the mail server which will host outgoing reports. The Simple Mail Transfer Protocol (SMTP) documentation for any mail server will contain information on the correct `port` to expose, typically `587`, and the name of the server. Replace `server`, `user`, and `password` fields with legitimate credentials.
````
Mail server will SEND messages from ATSD, not `host outgoing reports`.

kghamilton89 replied:
Rephrased.

rodionos commented on how-to/github/pr-report.md:
```diff
@@ -0,0 +1,173 @@
+docker run -d -p 8443:8443 \
+  --name=atsd-sandbox \
+  --env START_COLLECTOR=off \
+  --env SERVER_URL=https://atsd.company_name.com:8443 \
```
SERVER_URL is not necessary. We're not receiving any incoming webhooks from GutHub.

rodionos commented on how-to/github/pr-report.md:
```diff
@@ -0,0 +1,173 @@
+
+## Generating OAuth Access Token
+
+The GitHub API may be accessed [several ways](https://developer.github.com/v3/auth/#other-authentication-methods). This tool uses an [OAuth Personal Token](https://blog.github.com/2013-05-16-personal-api-tokens/) to query the API without sending user login information. Follow these instructions to generate a personal token and store it in the local ATSD instance.
```
Replace `tool` with `this integration`. Tool is often a term reserved for small programs like a text editor.

rodionos commented on how-to/github/pr-report.md:
````diff
@@ -0,0 +1,173 @@
+
+The bound volume should point to the **absolute path** where a plaintext file is stored containing the following parameters:
+
+```txt
````
I think we support sending a test messages to a user-defined address (similar to Slack test message) to verify that configuration is correct. Lets enable it here and display how this test message looks like.

kghamilton89 replied:
Added.

rodionos commented on how-to/github/pr-report.md:
````diff
@@ -0,0 +1,173 @@
+Modify delivery time by opening the `github-daily-pr-status` rule from the **Rules** page. The `Condition` field contains:
+
+```java
+now.getHourOfDay == 5
````
```javascript
// I think brackets are missing?
now.getHourOfDay()
```

kghamilton89 replied:
Brackets added.

rodionos commented on how-to/github/pr-report.md:
````diff
@@ -0,0 +1,173 @@
+
+Minute granularity may be applied by extending the expression:
+
+```java
````
Aren't brackets missing?

kghamilton89 replied:
And here.

rodionos commented on how-to/github/pr-report.md:
````diff
@@ -0,0 +1,173 @@
+now.getHourOfDay == 18 && now.getMinuteOfHour == 30
+```
+
+Report delivery will be scheduled for 6:30 PM UTC time.
````
Are you sure it will work? I think the `timer_1h` metric we use arrives only once an hour, so it will never arrive at the 30th minute.

kghamilton89 replied:
I changed the expression to `timer_1m` so I didn't have to wait so long to receive a test email which made it work. I'll just remove this setting I thought it was just something helpful to mention.

rodionos commented on how-to/github/pr-report.md:
```diff
@@ -0,0 +1,173 @@
+
+![](images/data-replacement-tables.png)
+
+The GraphQL query will be present in the `value` field. The `organization` clause contains the placeholder value `your-organization-name` surrounded by quotation marks. Without deleting the quotation marks, replace the template information with the case-sensitive name of the GitHub organization whose repositories will be monitored. For information about creating a new organization, see the [GitHu ...
```
Lets describe how the user can set the correct organization name and the delivery email in the docker run command using https://github.com/axibase/dockers/tree/atsd-sandbox#variable-substitution: `${ENV.ORG}` and `${ENV.EMAIL_RECEPIENT}`. This way the user doesn't have to open pages in ATSD in order to set delivery email and to modify organization.

kghamilton89 replied:
Variables added.

kghamilton89 commented:
* Included substitution variables with the launch command and simplified it, but still left instructions on adding additional email subscribers if needed after launch;
* Modified xml import files with sub variables;
* Removed instructions for defining Organization in GQL query, because it is now unneeded;
* Restructured **Report Delivery** conditions to reflect new launch commands;
* Made requested changes to all documents.

rodionos commented on how-to/github/fork-notification.md:
````diff
@@ -81,3 +81,13 @@ You'll begin receiving messenger notifications the next time someone creates a n
+The ATSD host will be present in the logs as a clickable link:
+
+```txt
+[ATSD] https://atsd_hostname:8443
````
When launched inside Docker, these links will not clickable because they're internal to the container where ATSD is launched. We could just say that the ATSD user interface is accessible on port 8443/https, or we could refer to Docker hostname as in https://github.com/axibase/dockers/tree/atsd-sandbox#exposed-ports

kghamilton89 replied:
Corrected and linked to docs.

rodionos commented on how-to/github/pr-report.md:
````diff
@@ -0,0 +1,162 @@
+  axibase/atsd-sandbox:latest
+```
+
+The `ORGANIZATION` variable should contain the case-sensitive name of the organization whose repositories should be queried when generating the daily report. The `TOKEN` variable should contain the [GitHub OAuth token](#generating-oauth-access-token) which must be generated to query repositories using the GraphQL API. The `SUBSCRIBERS` variable should contain the comma-separated list of email add ...
````
Move the ORG and SUBS variables before the command.

kghamilton89 replied:
Moved.

rodionos commented on how-to/github/pr-report.md:
```diff
@@ -0,0 +1,162 @@
+
+For information about creating a new organization, see the [GitHub Help Documentation](https://help.github.com/articles/creating-a-new-organization-from-scratch/).
+
+The bound volume should point to the **absolute path** where a plaintext file is stored containing the following parameters:
```
I would mention that because mail configuration involves quite a few parameters it makes sense to pass them via a file mounted into the container.

rodionos commented on how-to/github/pr-report.md:
````diff
@@ -0,0 +1,162 @@
+test_email=myuser@example.org
+```
+
+This file defines the mail server which will send outgoing reports. The Simple Mail Transfer Protocol (SMTP) documentation for any mail server will contain information on the correct `port` to expose, typically `587`, and the name of the server. Replace `server`, `user`, `password`, and `test_email` fields with legitimate credentials. The `test_email` parameter defines to whom ATSD will send a te ...
````
Lets reference https://github.com/axibase/atsd/blob/master/administration/mail-client.md in case the user needs to reconfigure the mail client after the launch.

kghamilton89 replied:
Add line on reconfiguration after launch.

rodionos commented on how-to/github/pr-report.md:
````diff
@@ -0,0 +1,162 @@
+The ATSD host will be present in the logs as a clickable link:
+
+```txt
+[ATSD] https://atsd_hostname:8443
````
It may not be accessible on the link because the atsd_hostname is known only to the container.

kghamilton89 replied:
Corrected and linked to docs.

rodionos commented on how-to/github/pr-report.md:
````diff
@@ -0,0 +1,162 @@
+now.getHourOfDay() == 5
+```
+
+Change the value of this expression to the integer UTC 24-hour time when the report should be delivered.
````
The time will be based on the server local time (our ATSD instances run on UTC time, but customers may have different setups).

kghamilton89 replied:
All time references changed to refer to server local time and not UTC time.

rodionos commented on how-to/github/pr-report.md:
```diff
@@ -0,0 +1,162 @@
+The `queryConfig` clause calls `github-graphql-table` which queries the [GraphQL API v4](https://developer.github.com/v4/guides/forming-calls/#the-graphql-endpoint) via POST method and returns open Pull Request information in JSON format.
+
+The `'GQL_query'` variable is delivered as the outgoing query and returns the `pullRequests` [node](https://developer.github.com/v4/guides/intro-to-graphql/#node), which is a JSON list of open Pull Requests.
+
```
Lets add the query text here. Note that I replaced axibase with apache and limited the query to 5 repos and 5 issues in each repo (we use 100 and 10).

```graphql
query { 
	organization(login: "apache") {
        repositories(first: 5, orderBy: {field: PUSHED_AT, direction: DESC}) {
        nodes {
            #name
            pullRequests(first: 5, states: OPEN, orderBy: {field: UPDATED_AT, direction: DESC}) {
                nodes {
                    headRepository { nameWithOwner }
                    url
                    author {
                    ... on User {
                        login name
                        }
                    }       
                    mergeable
                    baseRefName
                    headRefName
                    title
                    #milestone { title }
                    #labels(first: 3) { nodes{name} }
                    ... on PullRequest {
                        pullRequestcommits: commits(last: 1) {
                            #totalCount	
                            nodes {
                                commit {
                                    #url 
                                    status { state contexts { context description createdAt targetUrl } }
                                }
                            }
                        }
                    }
                }
            } 
        }      
    }
  }
}
```

kghamilton89 replied:
Query text added and explained, how to modify query parameters and links to GraphQL docs.

rodionos commented on how-to/github/pr-report.md:
````diff
@@ -0,0 +1,162 @@
+```
+
+Report delivery will be scheduled for 6:00 PM UTC time.
+
````
We need to instruct the user how to test the rule, because it will not fire until scheduled. Replacing timer_1h with timer_1m seems like a hack. This needs to be discussed, maybe under howto channel.

kghamilton89 replied:
Rule modified to sent report immediately on launch and then every day at 5 AM.

rodionos commented:
There are two main issues remaining with this article:

a) How to test a rule. Lets go with `timer_1m` so that the rule fires on start. Ensure however, that the rule will not fire more than 1 per day when ATSD is running.

b) How to exclude pull requests that are CONFLICTING or have failed checks. We could ask Anton to suggest a json path to including only mergeable PRs with successful status checks.

The mergeable PRs can be filtered as follows:

```javascript
$..pullRequests.nodes[?(@.mergeable == 'MERGEABLE')]
```

But I'm having a hard time figuring out how to include only those PRs where all status checks are OK:

```javascript
$..pullRequests.nodes[?(@.mergeable == 'MERGEABLE')][*][*][*][*][?(@.state == 'SUCCESS')]
```

I'm using http://jsonpath.com/ for testing.

```json
{
	"data": {
		"organization": {
			"repositories": {
				"nodes": [
					{
						"pullRequests": {
							"nodes": [
								{
									"headRepository": {
										"nameWithOwner": "axibase/atsd"
									},
									"url": "https://github.com/axibase/atsd/pull/582",
									"author": {
										"login": "annstriganova",
										"name": "Anna Striganova"
									},
									"mergeable": "MERGEABLE",
									"baseRefName": "master",
									"headRefName": "forecast",
									"title": "forecasting docs update",
									"pullRequestcommits": {
										"nodes": [
											{
												"commit": {
													"status": {
														"state": "PENDING",
														"contexts": [
															{
																"context": "continuous-integration/travis-ci/pr",
																"description": "The Travis CI build is in progress",
																"createdAt": "2018-05-14T15:30:00Z",
																"targetUrl": "https://travis-ci.org/axibase/atsd/builds/378779268?utm_source=github_status&utm_medium=notification"
															}
														]
													}
												}
											}
										]
									}
								}
							]
						}
					},
					{
						"pullRequests": {
							"nodes": [
								{
									"headRepository": {
										"nameWithOwner": "axibase/atsd-use-cases"
									},
									"url": "https://github.com/axibase/atsd-use-cases/pull/248",
									"author": {
										"login": "kghamilton89",
										"name": "KG Hamilton"
									},
									"mergeable": "MERGEABLE",
									"baseRefName": "master",
									"headRefName": "github-pr-report",
									"title": "GitHub pr report",
									"pullRequestcommits": {
										"nodes": [
											{
												"commit": {
													"status": {
														"state": "SUCCESS",
														"contexts": [
															{
																"context": "continuous-integration/travis-ci/pr",
																"description": "The Travis CI build passed",
																"createdAt": "2018-05-14T15:21:30Z",
																"targetUrl": "https://travis-ci.org/axibase/atsd-use-cases/builds/378774667?utm_source=github_status&utm_medium=notification"
															}
														]
													}
												}
											}
										]
									}
								},
								{
									"headRepository": {
										"nameWithOwner": "axibase/atsd-use-cases"
									},
									"url": "https://github.com/axibase/atsd-use-cases/pull/200",
									"author": {
										"login": "unrealwork",
										"name": "Igor Shmagrinskiy"
									},
									"mergeable": "MERGEABLE",
									"baseRefName": "master",
									"headRefName": "dev-howto-monitor-ssl-for-domains",
									"title": "[HowTo] How to Monitor SSL Certificates Expiry Dates for subdomains",
									"pullRequestcommits": {
										"nodes": [
											{
												"commit": {
													"status": {
														"state": "SUCCESS",
														"contexts": [
															{
																"context": "continuous-integration/travis-ci/pr",
																"description": "The Travis CI build passed",
																"createdAt": "2018-05-04T12:54:12Z",
																"targetUrl": "https://travis-ci.org/axibase/atsd-use-cases/builds/374873943?utm_source=github_status&utm_medium=notification"
															}
														]
													}
												}
											}
										]
									}
								}
							]
						}
					},
					{
						"pullRequests": {
							"nodes": []
						}
					},
					{
						"pullRequests": {
							"nodes": []
						}
					},
					{
						"pullRequests": {
							"nodes": []
						}
					},
					{
						"pullRequests": {
							"nodes": []
						}
					},
					{
						"pullRequests": {
							"nodes": []
						}
					},
					{
						"pullRequests": {
							"nodes": []
						}
					},
					{
						"pullRequests": {
							"nodes": []
						}
					},
					{
						"pullRequests": {
							"nodes": []
						}
					},
					{
						"pullRequests": {
							"nodes": [
								{
									"headRepository": {
										"nameWithOwner": "axibase/atsd-api-test"
									},
									"url": "https://github.com/axibase/atsd-api-test/pull/487",
									"author": {
										"login": "unrealwork",
										"name": "Igor Shmagrinskiy"
									},
									"mergeable": "MERGEABLE",
									"baseRefName": "master",
									"headRefName": "5208-series-tag-query-with-wildcard-without-entity",
									"title": "5208: Series tags query with wildcard without entity",
									"pullRequestcommits": {
										"nodes": [
											{
												"commit": {
													"status": {
														"state": "SUCCESS",
														"contexts": [
															{
																"context": "Codacy/PR Quality Review",
																"description": "Up to standards. A positive pull request.",
																"createdAt": "2018-05-06T20:36:16Z",
																"targetUrl": "https://app.codacy.com/app/unrealwork/atsd-api-test/pullRequest?prid=1550017"
															},
															{
																"context": "continuous-integration/travis-ci/pr",
																"description": "The Travis CI build passed",
																"createdAt": "2018-05-06T07:03:01Z",
																"targetUrl": "https://travis-ci.org/axibase/atsd-api-test/builds/375459238?utm_source=github_status&utm_medium=notification"
															}
														]
													}
												}
											}
										]
									}
								},
								{
									"headRepository": {
										"nameWithOwner": "axibase/atsd-api-test"
									},
									"url": "https://github.com/axibase/atsd-api-test/pull/465",
									"author": {
										"login": "alexey-kravtsov",
										"name": null
									},
									"mergeable": "MERGEABLE",
									"baseRefName": "master",
									"headRefName": "kravtsov-4814",
									"title": "SQL: Interpolation previous with fill=false #4814",
									"pullRequestcommits": {
										"nodes": [
											{
												"commit": {
													"status": {
														"state": "FAILURE",
														"contexts": [
															{
																"context": "continuous-integration/travis-ci/pr",
																"description": "The Travis CI build failed",
																"createdAt": "2018-05-06T07:02:09Z",
																"targetUrl": "https://travis-ci.org/axibase/atsd-api-test/builds/375459132?utm_source=github_status&utm_medium=notification"
															},
															{
																"context": "Codacy/PR Quality Review",
																"description": "Up to standards. A positive pull request.",
																"createdAt": "2018-05-06T20:38:55Z",
																"targetUrl": "https://app.codacy.com/app/unrealwork/atsd-api-test/pullRequest?prid=1257667"
															}
														]
													}
												}
											}
										]
									}
								},
								{
									"headRepository": {
										"nameWithOwner": "axibase/atsd-api-test"
									},
									"url": "https://github.com/axibase/atsd-api-test/pull/489",
									"author": {
										"login": "vtols",
										"name": "Valery Tolstov"
									},
									"mergeable": "MERGEABLE",
									"baseRefName": "master",
									"headRefName": "tolstov-5150",
									"title": "Short time format tests #5150",
									"pullRequestcommits": {
										"nodes": [
											{
												"commit": {
													"status": {
														"state": "FAILURE",
														"contexts": [
															{
																"context": "continuous-integration/travis-ci/pr",
																"description": "The Travis CI build failed",
																"createdAt": "2018-04-25T13:18:10Z",
																"targetUrl": "https://travis-ci.org/axibase/atsd-api-test/builds/371056008?utm_source=github_status&utm_medium=notification"
															},
															{
																"context": "Codacy/PR Quality Review",
																"description": "Up to standards. A positive pull request.",
																"createdAt": "2018-04-25T16:14:56Z",
																"targetUrl": "https://app.codacy.com/app/unrealwork/atsd-api-test/pullRequest?prid=1573849"
															}
														]
													}
												}
											}
										]
									}
								},
								{
									"headRepository": {
										"nameWithOwner": "axibase/atsd-api-test"
									},
									"url": "https://github.com/axibase/atsd-api-test/pull/488",
									"author": {
										"login": "raipc",
										"name": "Anton Rybochkin"
									},
									"mergeable": "MERGEABLE",
									"baseRefName": "master",
									"headRefName": "revert-raipc",
									"title": "revert debug changes",
									"pullRequestcommits": {
										"nodes": [
											{
												"commit": {
													"status": {
														"state": "FAILURE",
														"contexts": [
															{
																"context": "continuous-integration/travis-ci/pr",
																"description": "The Travis CI build failed",
																"createdAt": "2018-04-20T10:33:38Z",
																"targetUrl": "https://travis-ci.org/axibase/atsd-api-test/builds/369065656?utm_source=github_status&utm_medium=notification"
															},
															{
																"context": "Codacy/PR Quality Review",
																"description": "Up to standards. A positive pull request.",
																"createdAt": "2018-04-20T12:03:05Z",
																"targetUrl": "https://app.codacy.com/app/unrealwork/atsd-api-test/pullRequest?prid=1561823"
															}
														]
													}
												}
											}
										]
									}
								},
								{
									"headRepository": {
										"nameWithOwner": "axibase/atsd-api-test"
									},
									"url": "https://github.com/axibase/atsd-api-test/pull/486",
									"author": {
										"login": "unrealwork",
										"name": "Igor Shmagrinskiy"
									},
									"mergeable": "MERGEABLE",
									"baseRefName": "master",
									"headRefName": "disable_large_test",
									"title": "Revert \"SQL: Сolumn name sensitivity in subqueries #5139 (#480)\"",
									"pullRequestcommits": {
										"nodes": [
											{
												"commit": {
													"status": {
														"state": "FAILURE",
														"contexts": [
															{
																"context": "continuous-integration/travis-ci/pr",
																"description": "The Travis CI build failed",
																"createdAt": "2018-04-16T08:24:30Z",
																"targetUrl": "https://travis-ci.org/axibase/atsd-api-test/builds/367039454?utm_source=github_status&utm_medium=notification"
															},
															{
																"context": "Codacy/PR Quality Review",
																"description": "Up to standards. A positive pull request.",
																"createdAt": "2018-04-16T23:52:26Z",
																"targetUrl": "https://app.codacy.com/app/unrealwork/atsd-api-test/pullRequest?prid=1510635"
															}
														]
													}
												}
											}
										]
									}
								},
								{
									"headRepository": {
										"nameWithOwner": "axibase/atsd-api-test"
									},
									"url": "https://github.com/axibase/atsd-api-test/pull/483",
									"author": {
										"login": "unrealwork",
										"name": "Igor Shmagrinskiy"
									},
									"mergeable": "MERGEABLE",
									"baseRefName": "master",
									"headRefName": "disable_authentication",
									"title": "Disable authentication",
									"pullRequestcommits": {
										"nodes": [
											{
												"commit": {
													"status": {
														"state": "FAILURE",
														"contexts": [
															{
																"context": "continuous-integration/travis-ci/pr",
																"description": "The Travis CI build failed",
																"createdAt": "2018-04-04T16:48:37Z",
																"targetUrl": "https://travis-ci.org/axibase/atsd-api-test/builds/362225447?utm_source=github_status&utm_medium=notification"
															},
															{
																"context": "Codacy/PR Quality Review",
																"description": "Good work! A positive pull request.",
																"createdAt": "2018-04-04T16:43:47Z",
																"targetUrl": "https://app.codacy.com/app/unrealwork/atsd-api-test/pullRequest?prid=1505551"
															}
														]
													}
												}
											}
										]
									}
								},
								{
									"headRepository": {
										"nameWithOwner": "axibase/atsd-api-test"
									},
									"url": "https://github.com/axibase/atsd-api-test/pull/406",
									"author": {
										"login": "vtols",
										"name": "Valery Tolstov"
									},
									"mergeable": "MERGEABLE",
									"baseRefName": "master",
									"headRefName": "tolstov-4397",
									"title": "Test #4397",
									"pullRequestcommits": {
										"nodes": [
											{
												"commit": {
													"status": {
														"state": "FAILURE",
														"contexts": [
															{
																"context": "continuous-integration/travis-ci/pr",
																"description": "The Travis CI build failed",
																"createdAt": "2017-12-14T17:56:19Z",
																"targetUrl": "https://travis-ci.org/axibase/atsd-api-test/builds/316544214?utm_source=github_status&utm_medium=notification"
															}
														]
													}
												}
											}
										]
									}
								}
							]
						}
					},
					{
						"pullRequests": {
							"nodes": []
						}
					},
					{
						"pullRequests": {
							"nodes": []
						}
					},
					{
						"pullRequests": {
							"nodes": []
						}
					},
					{
						"pullRequests": {
							"nodes": []
						}
					},
					{
						"pullRequests": {
							"nodes": []
						}
					},
					{
						"pullRequests": {
							"nodes": []
						}
					},
					{
						"pullRequests": {
							"nodes": []
						}
					},
					{
						"pullRequests": {
							"nodes": [
								{
									"headRepository": {
										"nameWithOwner": "axibase/atsd-jdbc-test"
									},
									"url": "https://github.com/axibase/atsd-jdbc-test/pull/6",
									"author": {
										"login": "raipc",
										"name": "Anton Rybochkin"
									},
									"mergeable": "CONFLICTING",
									"baseRefName": "master",
									"headRefName": "atsd-jdbc-migration",
									"title": "Migrated tests from atsd-jdbc [WIP]",
									"pullRequestcommits": {
										"nodes": [
											{
												"commit": {
													"status": {
														"state": "SUCCESS",
														"contexts": [
															{
																"context": "codacy/pr",
																"description": "Good work! A positive pull request.",
																"createdAt": "2017-08-09T01:43:11Z",
																"targetUrl": "https://www.codacy.com/app/anton-rib/atsd-jdbc-test/pullRequest?prid=828794"
															}
														]
													}
												}
											}
										]
									}
								}
							]
						}
					},
					{
						"pullRequests": {
							"nodes": []
						}
					},
					{
						"pullRequests": {
							"nodes": []
						}
					},
					{
						"pullRequests": {
							"nodes": []
						}
					},
					{
						"pullRequests": {
							"nodes": []
						}
					},
					{
						"pullRequests": {
							"nodes": []
						}
					},
					{
						"pullRequests": {
							"nodes": []
						}
					},
					{
						"pullRequests": {
							"nodes": []
						}
					},
					{
						"pullRequests": {
							"nodes": []
						}
					},
					{
						"pullRequests": {
							"nodes": []
						}
					},
					{
						"pullRequests": {
							"nodes": []
						}
					},
					{
						"pullRequests": {
							"nodes": []
						}
					},
					{
						"pullRequests": {
							"nodes": []
						}
					},
					{
						"pullRequests": {
							"nodes": []
						}
					},
					{
						"pullRequests": {
							"nodes": []
						}
					},
					{
						"pullRequests": {
							"nodes": []
						}
					},
					{
						"pullRequests": {
							"nodes": []
						}
					},
					{
						"pullRequests": {
							"nodes": []
						}
					},
					{
						"pullRequests": {
							"nodes": []
						}
					},
					{
						"pullRequests": {
							"nodes": []
						}
					},
					{
						"pullRequests": {
							"nodes": []
						}
					},
					{
						"pullRequests": {
							"nodes": []
						}
					},
					{
						"pullRequests": {
							"nodes": []
						}
					},
					{
						"pullRequests": {
							"nodes": []
						}
					},
					{
						"pullRequests": {
							"nodes": []
						}
					},
					{
						"pullRequests": {
							"nodes": []
						}
					},
					{
						"pullRequests": {
							"nodes": []
						}
					},
					{
						"pullRequests": {
							"nodes": []
						}
					},
					{
						"pullRequests": {
							"nodes": []
						}
					},
					{
						"pullRequests": {
							"nodes": []
						}
					},
					{
						"pullRequests": {
							"nodes": []
						}
					},
					{
						"pullRequests": {
							"nodes": []
						}
					},
					{
						"pullRequests": {
							"nodes": []
						}
					},
					{
						"pullRequests": {
							"nodes": []
						}
					}
				]
			}
		}
	}
}
```

rodionos commented:
Here's the JSON Path expression that selects only mergeable PRs with passed Status checks:

```javascript
$..pullRequests.nodes[?(@.mergeable == 'MERGEABLE' && @.pullRequestcommits.nodes[0].commit.status.state == 'SUCCESS')]
```

It looks like we need to create two rules (two reports): 'PRs ready to merge' and 'PRs failing'. We can use the same rule, just add an extra email notification. Two reports will be sent at once.

rodionos commented:
I found a good example for the section where we discuss functions used in the email notification: https://developer.github.com/v4/guides/forming-calls/#example-query. It's a step by step review where results of each step are explained.

kghamilton89 commented:
Latest push:

* Expand **Overview** to include new scope, PR reports for both `SUCCESS` and `FAILURE` status.
* Modify email notification for rule to include `SUCCESS` and `FAILURE` reports.
* Attach GitHub Dev Docs ahead of GraphQL query.

kghamilton89 commented:
`test-email` parameter removed and text refactored to reflect automatically generated test email.

rodionos commented on how-to/github/pr-report.md:
````diff
@@ -0,0 +1,210 @@
+```txt
+server_name=ATSD-sandbox
+server=smtp.example.org
+port=587
````
Only 3 mail properties are now required: https://github.com/axibase/dockers/tree/atsd-sandbox#usage-example

Retest only with required Mail Client settings to simplify the configuration. Add `test_email` to mail.properties with the same value as SUBSCRIBERS.

Tell the user what to examine in the log file in order to verify that email is working? If OK, they should receive a test email to `SUBSCRIBERS`. If not, what will be displayed in log file?

kghamilton89 replied:
Log output added and parameters updated.

rodionos commented on how-to/github/pr-report.md:
````diff
@@ -0,0 +1,210 @@
+```
+
+This file defines the mail server which will send outgoing reports. The Simple Mail Transfer Protocol (SMTP) documentation for any mail server will contain information on the correct `port` to expose, typically `587`, and the name of the server. Replace `server`, `user`, and `password` fields with legitimate credentials. Upon launch, the **Mail Client** will automatically send the user defined in ...
+
````
Because we will manually set `test_email` property in the file, the test email will be delivered to a different mailbox. Namely to the address set in `test_email`. It should be the same as `SUBSCRIBERS` variable.

kghamilton89 replied:
Added.

rodionos commented on how-to/github/pr-report.md:
````diff
@@ -0,0 +1,210 @@
+```sh
+docker logs -f atsd-sandbox
+```
+
````
We also need to check that email was configured properly. This will be displayed in the log file.

kghamilton89 replied:
Mentioned passing and failing log output.

rodionos commented on how-to/github/pr-report.md:
```diff
@@ -0,0 +1,210 @@
+
+![](images/read-org.png)
+
+Configure the token to grant **read:org** permissions in the **admin:org** section. This scope grants read-only organization access to any user with this token, keep it confidential. For more information about token scopes, see [GitHub Developer Documentation](https://developer.github.com/apps/building-oauth-apps/scopes-for-oauth-apps/).
```
The screenshot shows a different permission. We only need `read:org`. Re-test with `read:org`.  You chose `Admin read public key` - why?

rodionos commented on how-to/github/pr-report.md:
```diff
@@ -0,0 +1,210 @@
+Configure the token to grant **read:org** permissions in the **admin:org** section. This scope grants read-only organization access to any user with this token, keep it confidential. For more information about token scopes, see [GitHub Developer Documentation](https://developer.github.com/apps/building-oauth-apps/scopes-for-oauth-apps/).
+
+## Adding Additional Subscribers
+
```
Remove this section as non-essential.

kghamilton89 replied:
Removed.

rodionos commented on how-to/github/pr-report.md:
```diff
@@ -0,0 +1,210 @@
+
+Open the **Settings** menu and select **System Information** to view server local time.
+
+![](images/settings-system-information.png)
```
Remove System Information screenshot - this is not essential.

kghamilton89 replied:
Removed.

rodionos commented on how-to/github/pr-report.md:
````diff
@@ -0,0 +1,210 @@
+```
+
+ATSD web interface is accessible at [`https://docker_host:8443/`](https://github.com/axibase/dockers/tree/atsd-sandbox#exposed-ports).
+
````
Add a section on how to get the report for the first time, ahead of schedule. There is no easy way - so we need to discuss it.

kghamilton89 replied:
The way the rule is written now, the first report will be delivered when the instance is started.

rodionos commented on how-to/github/pr-report.md:
````diff
@@ -0,0 +1,210 @@
+
+Consider this query:
+
+```javascript
````
this is a GraphQL query, not `javascript`.

There should be a specific syntax for `graphql`.

kghamilton89 replied:
Replaced.

rodionos commented on how-to/github/pr-report.md:
```diff
@@ -0,0 +1,210 @@
+This query targets [The Apache Software Foundation](https://github.com/apache) repositories and returns a daily report with the first five Pull Requests from the first five repositories in their GitHub library, alphabetically. These settings may be configured by modifying the `pullRequests(first: 5, states: OPEN, orderBy: {field: UPDATED_AT, direction: DESC}) {` and `repositories(first:5, orderBy ...
+
+For additional GraphQL query syntax, view the [documentation](https://graphql.org/learn/queries/).
+
```
OK, we execute this graphQL query. What's the output?
Then we filter the GraphQL query with JsonPath. What's the output after path?
Finally, we pass the json after filtering into add addTable function. What's the output?

Please add links to rule engine functions that we're using.

kghamilton89 replied:
Step-by-step GraphQL process added, outputs, JSONPaths and filtered results added as well as Rule Engine links and real reports from Apache repositories.
